math/minuit2/CMakeLists.txt

```diff
@@ -4,23 +4,62 @@
 
 add_definitions(-DWARNINGMSG -DUSE_ROOT_ERROR)
 
-#---Deal with the parallel option on Minuit2. Probably it should be done using a build 'option' and not
-#   using a environment variable  -- NOT TESTED ---
-if($ENV{USE_PARALLEL_MINUIT2})
-  if($ENV{USE_OPENMP})
-    add_definitions(-D_GLIBCXX_PARALLEL -fopenmp)
-    set_target_properties(Minuit2 PROPERTIES LINK_FLAGS -fopenmp)
-  elseif($ENV{USE_MPI})
-    add_definitions(-DMPIPROC)
-    set(CMAKE_CXX_COMPILER mpic++)
-    set(CMAKE_C_COMPILER mpic++)
-    set(CMAKE_CXX_LINK_EXECUTABLE mpic++)
-  endif()
-endif()
+#---Deal with the parallel option on Minuit2.
+#   -- NOT TESTED ---
+option(MINUIT2_OMP "Use OpenMP for Minuit2 (not tested)")
+option(MINUIT2_MPI "Use MPI for Minuit2 (not tested)")
+
 
 ROOT_STANDARD_LIBRARY_PACKAGE(Minuit2
                               HEADERS *.h Minuit2/*.h
                               DICTIONARY_OPTIONS "-writeEmptyRootPCM"
                               DEPENDENCIES MathCore Hist)
 
+if(MINUIT2_OMP)
+    find_package(OpenMP REQUIRED)
+    add_definitions(-D_GLIBCXX_PARALLEL)
+
+    # For CMake < 3.9, we need to make the target ourselves
+    if(NOT TARGET OpenMP::OpenMP_CXX)
+        add_library(OpenMP::OpenMP_CXX IMPORTED INTERFACE)
+        set_property(TARGET OpenMP::OpenMP_CXX
+                     PROPERTY INTERFACE_COMPILE_OPTIONS ${OpenMP_CXX_FLAGS})
+        # Only works if the same flag is passed to the linker; use CMake 3.9+ otherwise (Intel, AppleClang)
+        set_property(TARGET OpenMP::OpenMP_CXX
+                     PROPERTY INTERFACE_LINK_LIBRARIES ${OpenMP_CXX_FLAGS})
+
+
+        set(THREADS_PREFER_PTHREAD_FLAG ON)
+        find_package(Threads REQUIRED)
+        target_link_libraries(Minuit2 INTERFACE Threads::Threads)
+    endif()
+    target_link_libraries(Minuit2 OpenMP::OpenMP_CXX)
+
+    message(STATUS "Minuit2: Building with  with OpenMP support")
+endif()
+
+if(MINUIT2_MPI)
+    find_package(MPI REQUIRED)
+    add_definitions(-DMPIPROC)
+
+    if(NOT TARGET MPI::MPI_CXX)
+        add_library(MPI::MPI_CXX IMPORTED INTERFACE)
+
+        set_property(TARGET MPI::MPI_CXX
+                     PROPERTY INTERFACE_COMPILE_OPTIONS ${MPI_CXX_COMPILE_FLAGS})
+        set_property(TARGET MPI::MPI_CXX
+                     PROPERTY INTERFACE_INCLUDE_DIRECTORIES "${MPI_CXX_INCLUDE_PATH}")
+        set_property(TARGET MPI::MPI_CXX
+                     PROPERTY INTERFACE_LINK_LIBRARIES ${MPI_CXX_LINK_FLAGS} ${MPI_CXX_LIBRARIES})
+    endif()
+
+    target_link_libraries(Minuit2 MPI::MPI_CXX)
+    message(STATUS "Minuit2: Using MPI bindings")
+    message(STATUS "Minuit2: Run: ${MPIEXEC} ${MPIEXEC_NUMPROC_FLAG} ${MPIEXEC_MAX_NUMPROCS} ${MPIEXEC_PREFLAGS} EXECUTABLE ${MPIEXEC_POSTFLAGS} ARGS")
+
+    # set(CMAKE_CXX_COMPILER mpic++)
+    # set(CMAKE_C_COMPILER mpic++)
+    # set(CMAKE_CXX_LINK_EXECUTABLE mpic++)
+endif()
```

Comment:
I've already made very similar changes to this locally, so that we can use this for both ROOT and the standalone Minuit2. Once I merge, I will test and merge the previous commit from this PR. Please wait a bit until I try to fix the broken test, then we can move forward with this.

Reply:
Okay. I have a general idea of how to integrate the two CMake's once this is done.